File: baekjoon-Algorithm/algorithm-middle/brute-force/N9663.py
import sys

input = sys.stdin.readline

N = int(input())

cnt = 0
row = [0] * N


def is_promising(x):
    for i in range(x):
        # 세로 체크, 대각선 체크
        if row[x] == row[i] or abs(row[x] - row[i]) == x - i:
            return False

    return True


def n_queens(x):
    global cnt

    if x == N:
        cnt += 1
        return

    for i in range(N):
        row[x] = i  # (x, i)에 퀸을 놓음
        # 유망성 판단: 다음 칸에 놓을 수 있는 경우가 있는지
        if is_promising(x):
            n_queens(x + 1)


n_queens(0)
print(cnt)

File: baekjoon-Algorithm/algorithm-middle/brute-force/N9663.java
import java.io.*;

public class N9663 {
    static int N;
    static int[] row;
    static int cnt;

    static boolean is_promising(int x) {
        for (int i = 0; i < x; i++) {
            if (Math.abs(row[x] - row[i]) == x - i || row[x] == row[i]) {
                return false;
            }
        }
        return true;
    }

    static void n_queen(int x) {
        if (x == N) {
            cnt++;
            return;
        }

        for (int i = 0; i < N; i++) {
            row[x] = i;

            if (is_promising(x)) {
                n_queen(x+1);
            }
        }
    }

    public static void main (String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        row = new int[N];
        cnt = 0;

        n_queen(0);
        System.out.println(cnt);

    }
}
